src/bidi_class.rs

```diff
@@ -0,0 +1,145 @@
+use error::Result;
+use util::{print_property_values, PropertyValues};
+
+pub fn command(args: ArgMatches) -> Result<()> {
+    let dir = args.ucd_dir()?;
+    let propvals = PropertyValues::from_ucd_dir(&dir)?;
+    let rows: Vec<UnicodeData> = ucd_parse::parse(&dir)?;
+    let core_prop: Vec<CoreProperty> = ucd_parse::parse(&dir)?;
+    let use_short_names = args.is_present("short-names");
+    let bidi_class_name = |short_name: &str| {
+        if use_short_names {
+            Ok(short_name.to_string())
+        } else {
+            propvals.canonical("bc", short_name)
+        }
+    };
+
+    // If we were tasked with listing the available categories, then do that
+    // and quit.
+    if args.is_present("list-classes") {
+        return print_property_values(&propvals, "Bidi_Class");
+    }
+
+    // Collect each bidi class into an ordered set.
+    let mut by_type: BTreeMap<String, BTreeSet<u32>> = BTreeMap::new();
+    let mut assigned = BTreeSet::new();
+    for row in rows {
+        assigned.insert(row.codepoint.value());
+        let bc = bidi_class_name(&row.bidi_class)?;
+        by_type
+            .entry(bc)
+            .or_insert(BTreeSet::new())
+            .insert(row.codepoint.value());
+    }
+
+    // Process the codepoints that are not listed as per the notes in
+    // DerivedBidiClass.txt (UCD 12.1):
+    //
+    // Bidi Class (listing UnicodeData.txt, field 4: see UAX #44: http://www.unicode.org/reports/tr44/)
+    // Unlike other properties, unassigned code points in blocks
+    // reserved for right-to-left scripts are given either types R or AL.
+    //
+    // The unassigned code points that default to AL are in the ranges:
+    //     [\u0600-\u07BF \u0860-\u086F \u08A0-\u08FF \uFB50-\uFDCF \uFDF0-\uFDFF \uFE70-\uFEFF
+    //      \U00010D00-\U00010D3F \U00010F30-\U00010F6F
+    //      \U0001EC70-\U0001ECBF \U0001ED00-\U0001ED4F \U0001EE00-\U0001EEFF]
+    //
+    //     This includes code points in the Arabic, Syriac, and Thaana blocks, among others.
+    //
+    // The unassigned code points that default to R are in the ranges:
+    //     [\u0590-\u05FF \u07C0-\u085F \u0870-\u089F \uFB1D-\uFB4F
+    //      \U00010800-\U00010CFF \U00010D40-\U00010F2F \U00010F70-\U00010FFF
+    //      \U0001E800-\U0001EC6F \U0001ECC0-\U0001ECFF \U0001ED50-\U0001EDFF \U0001EF00-\U0001EFFF]
+    //
+    //     This includes code points in the Hebrew, NKo, and Phoenician blocks, among others.
+    //
+    // The unassigned code points that default to ET are in the range:
+    //     [\u20A0-\u20CF]
+    //
+    //     This consists of code points in the Currency Symbols block.
+    //
+    // The unassigned code points that default to BN have one of the following properties:
+    //     Default_Ignorable_Code_Point
+    //     Noncharacter_Code_Point
+    //
+    // For all other cases:
+    //
+    //  All code points not explicitly listed for Bidi_Class
+    //  have the value Left_To_Right (L).
+    #[rustfmt::skip]
```

Comment:
Hmmm... Could you say more about this? I'd really rather we stuck to rustfmt unless there's a really good reason. In particular, for this code, it exceeds my normal 79 column limit.

Reply:
Basically my thinking was along these lines:

* The comment was copied directly from `DerivedBidiClass.txt`. I didn't reformat it so that future updates from the spec would should only what changed
* `default_class_assignments` was the code the represented the data contained in the comment/spec. I tried to format/group the data to match the layout in the spec to make it easier to eyeball that the two were the same. I figured this might also be useful if the need arose to update it in the future. I added the `rustfmt::skip` to avoid rustfmt messing with that formatting.

Anyway I've moved it to a constant because once you pointed it out I'm not really sure what I was thinking sticking it inline like that. Since the comment and data are tied to each other I'm wasn't sure this would work will inside `lookup_unassigned` so I put it at the top level.